src/Serialization/containers.jl

```diff
@@ -25,15 +35,52 @@ function has_empty_entries(obj::T) where T <: ContainerTypes
   return false
 end
 
-function type_params(obj::S) where {T <: ContainerTypes, S <:MatVecType{T}}
+function type_params(obj::S) where {T <: ContainerTypes, S <: MatVecType{T}}
   isempty(obj) && return TypeParams(S, TypeParams(T, nothing))
 
   # empty entries can inherit params from the rest of the collection
   params = type_params.(filter(!has_empty_entries, obj))
-  @req params_all_equal(params) "Not all params of Vector or Matrix entries are the same, consider using a Tuple for serialization"
+  @req params_all_equal(params) "Not all params of the entries are the same, consider using a Tuple for serialization"
   return TypeParams(S, params[1])
 end
 
+function save_type_params(s::SerializerState,
+                          tp::TypeParams{T, <:Tuple{Vararg{Pair}}}) where {S, N,  T <: Array{S, N}}
+  save_data_dict(s) do
+    save_object(s, encode_type(T), :name)
+    save_data_dict(s, :params) do
+      for (k, v) in params(tp)
+        if k == :dims
+          save_object(s, v, k)
+          continue
+        end
+        save_type_params(s, v, k)
+      end
+    end
+  end
+end
+
+# this function is forced to deal with all array types
+function load_type_params(s::DeserializerState, T::Type{<: Union{Array, MatVecType}})
+  !haskey(s, :params) && return T, nothing
+  subtype, params = load_node(s, :params) do _
+    if haskey(s, :dims)
+      U = load_node(s, :subtype_params) do _
+        return decode_type(s)
+      end
+       
+      dims = load_object(s, Int, :dims)
+      subtype_without_dims, params =  load_type_params(s, U, :subtype_params)
+      subtype = subtype_without_dims, dims
+      return T{subtype...}, params
```

Comment:
```suggestion
      subtype, params =  load_type_params(s, U, :subtype_params)
      return T{subtype, dims}, params
```
this looks a bit clearer to me